presto-raptor/src/main/java/com/facebook/presto/raptor/storage/DeltaShardRewriter.java

```diff
@@ -0,0 +1,159 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.raptor.storage;
+
+import com.facebook.airlift.json.JsonCodec;
+import com.facebook.presto.orc.cache.OrcFileTailSource;
+import com.facebook.presto.raptor.filesystem.FileSystemContext;
+import com.facebook.presto.raptor.metadata.ShardDeleteDelta;
+import com.facebook.presto.raptor.metadata.ShardInfo;
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.BlockBuilder;
+import com.facebook.presto.spi.block.LongArrayBlockBuilder;
+import com.facebook.presto.spi.type.Type;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import io.airlift.slice.Slice;
+import io.airlift.slice.Slices;
+import org.apache.hadoop.fs.FileSystem;
+
+import java.util.BitSet;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+
+import static com.facebook.airlift.concurrent.MoreFutures.getFutureValue;
+import static com.facebook.airlift.json.JsonCodec.jsonCodec;
+import static com.facebook.presto.raptor.RaptorErrorCode.RAPTOR_ERROR;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static java.util.concurrent.CompletableFuture.supplyAsync;
+
+public class DeltaShardRewriter
+        implements ShardRewriter
+{
+    private static final JsonCodec<ShardDeleteDelta> SHARD_DELETE_DELTA_CODEC = jsonCodec(ShardDeleteDelta.class);
+    private final OrcStorageManager orcStorageManager;
+    private final FileSystemContext fileSystemContext;
+    private final FileSystem fileSystem;
+    private final OrcFileTailSource orcFileTailSource;
+    private final long transactionId;
+    private final OptionalInt bucketNumber;
+    private final UUID oldShardUuid;
+    private final int oldShardRowCount;
+    private final Optional<UUID> oldDeltaDeleteShardUuid;
+    private final ExecutorService deletionExecutor;
+    private final ReaderAttributes defaultReaderAttributes;
+
+    public DeltaShardRewriter(
+            FileSystemContext fileSystemContext,
+            FileSystem fileSystem,
```

Comment:
As my previous comments on `OrcPageSource` in the last commit, I don't think we need `FileSystemContext`, `FileSystem`, `BucketNumber`, `TransactionId` or `OrcStorageManager`. `DeltaShardRewriter` needs to know:
(1) How to get a `BitSet` from `UUID`; and (2) How to get a page sink to write delta. We can pass functions to this rewriter from `OrcStorageManager` when constructing the rewriter.